forwardrateagreement constructor enhancements fixes #154

jakeheke75 opened the pull request:
Hi @lballabio I tried to enhance a bit the constructors of `ForwardRateAgreement` class.
As I am a newbie c++ developer I look forward to your expert review/comments.
I tested some cases for the old and new constructors and the results seem consistent, I attach the code if the test file.
Any suggestion/improvement is welcome, also from @tomwhoiscontrary who actively participated in issue #154.
Thank you in advance for your time.

[FRA_test.cpp.txt](https://github.com/lballabio/QuantLib/files/10715026/FRA_test.cpp.txt)

jakeheke75 commented:
Mhh @lballabio I see that almost all the builds failed because of -Werror=reorder. Do I have to change the order of the initializer list in order to match the same members declaration order in the definition of class `ForwardRateAgreement` right? Thanks and sorry for the rookie error

lballabio commented:
That's right — no problem.

tomwhoiscontrary commented on ql/instruments/forwardrateagreement.cpp:
```diff
@@ -61,8 +61,72 @@ namespace QuantLib {
-        return index_->fixingDate(valueDate_);
+        if (index_)
+            return index_->fixingDate(valueDate_);
+        else // if FRA is constructed without the index it returns the valueDate
```
This seems dubious. If i create two FRAs, one without an index, and one without an index but with the exact same value and maturity dates as the one with an index, i would expect them to behave identically. But because of this, they won't - they will have different fixing dates.

jakeheke75 replied:
I agree. I decided to return the `valueDate` here because in case the FRA is created without the index then basically the fixing days can't be implicitly taken from it. Hence I imagined that the `valueDate` should already consider the number of fixing days. If we want to go for the solution where we either take the `IborIndex` or the `Handle<YieldTermStructure>` then I will add one parameter `fixingDays`in the constructor and I will calculate the `fixingDate` accordingly as `valueDate - fixingDays`.

tomwhoiscontrary commented on ql/instruments/forwardrateagreement.cpp:
```diff
@@ -61,8 +61,72 @@ namespace QuantLib {
     : ForwardRateAgreement(valueDate, index->maturityDate(valueDate), type, strikeForwardRate,
                            notionalAmount, index, std::move(discountCurve), true) {}
 
+    ForwardRateAgreement::ForwardRateAgreement(Position::Type type,
```
Type should come after value date for consistency with other constructors. I think Luigi has already suggested modifying the existing constructor rather than adding this one, which would fix this.

jakeheke75 replied:
Yep. Will update this.

tomwhoiscontrary commented on ql/instruments/forwardrateagreement.cpp:
```diff
@@ -61,8 +61,72 @@ namespace QuantLib {
+                                               const Date& maturityDate,
+                                               Position::Type type,
+                                               Rate strikeForwardRate,
+                                               Rate referenceRate,
```
Ah, i think this is the wrong approach. For the case with custom dates, we still want to get the rate from a YieldTermStructure. If i understand correctly, here you are requiring a fixed rate be supplied when the object is constructed.

In general, the way QuantLib works is that you define a graph of objects, where instruments depend on indexes, which depend on term structures, which depend on helpers, which depend on quotes. Then, you can update the quotes with some market data, and read out the prices for instruments implied by the market data. If you require a fixed rate like this, you make it impossible to construct that graph.

If you want to take Luigi's suggested approach of not taking an index, then i think you should take a `Handle<YieldTermStructure>` instead.

tomwhoiscontrary commented on ql/instruments/forwardrateagreement.cpp:
```diff
@@ -88,24 +152,36 @@ namespace QuantLib {
-        NPV_ = amount_ * discount->discount(valueDate_);
+            NPV_ = amount_ * discount->discount(valueDate_);
+        } else
+            NPV_ = amount_ * referenceRate_.discountFactor(valueDateTime_);
```
This is surely very unlikely to be right. The discount factor here describes how much a unit of currency at the FRA's value date is worth at the reference rate (today). But your reference rate applies to the term of the FRA (value date to maturity date), doesn't it? You can't just take a rate for the latter period and use it for the former period.

jakeheke75 replied:
In my intention here with the `referenceRate` I was meaning the index interest rate the FRA contract rate (`strikeForwardRate`) is compared against in order to determine the settlement amount. This comparison happens at the settlement date. The FRA effectively ends with the settlement date, as there is no longer any contractual engagement between the two counter parties once the settlement amount has been paid. The contract period is merely one of the calculation parameters used to determine the settlement amount. So if for instance we have a 1x3 FRA where `evaluationDate` is 31 January 2023, `valueDate` is 28 February 2023, `maturityDate` is 31 May 2023 then the Present Value at the `evaluationDate` is the FRA amount (the future cash flow) calculated at the `valueDate` multiplied by the discount factor. Anyway, as we said if the index is not provided then the `referenceRate`will be taken form the `Handle<YieldTermStructure>`.

tomwhoiscontrary reviewed (commented):
I am concerned that this is not the right approach. Also some small nits.

jakeheke75 commented on ql/instruments/forwardrateagreement.hpp:
```diff
@@ -82,6 +91,35 @@ namespace QuantLib {
+            \details constructor taking the index, the FRA maturityDate is calculated from the index
+                     in this way we are avoiding conflicting information in the definition of the FRA
+        */
+        ForwardRateAgreement(
```
Good point. Will do it, thanks.

jakeheke75 commented:
Hi @lballabio I fixed the wrong initializer list and created the new constructor taking the `Handle<YieldTermStructure>` instead of the index.  I built the whole library in Linux with CMake and executed the complete test-suite. Please let me know if we need to change something, thanks. I attach the c++ file used to test the behavior of class `ForwardRateAgreement.
`
[FRA_test.cpp.txt](https://github.com/lballabio/QuantLib/files/10777182/FRA_test.cpp.txt)

sweemer commented:
@jakeheke75 You should set `CMAKE_CXX_FLAGS="-Wall -Werror"` when building locally with CMake so that you get the same warnings and errors as in the CI.

jakeheke75 commented:
> @jakeheke75 You should set `CMAKE_CXX_FLAGS="-Wall -Werror"` when building locally with CMake so that you get the same warnings and errors as in the CI.

Thanks @sweemer will do it!

jakeheke75 commented:
> Apologies, busy period. I'll have a look.

Ciao Luigi, no worries. It is currently failing the test suite, specifically `piecewiseyieldcurve.cpp`. It is failing the member function `testCurveConsistency`. The job which is failing is the cmake-win-with-options, therefore the library has been compiled with the flag `QL_USE_INDEXED_COUPON` defined. It means that in the source code then we have `bool useIndexedFra = false`. The piece of code failing is the checking of the FRA consistency. What I modified here is the constructor call to `ForwardRateAgreement`: I removed the manual calculation of maturity date `Date end` and invoked the constructor without it. But... with the `bool useIndexedFra = true` (therefore with `QL_USE_INDEXED_COUPON` undefined) it is working like a charm.
This is the case where the FRA maturity date is taken from the `Euribor3M` index `fixingDate`. In this case the `forwardRate` is calculated from the index `index_->fixing(fixingDate())`.
Instead with the `bool useIndexedFra = false` (therefore with `QL_USE_INDEXED_COUPON` defined) it is the case where the FRA maturity date is taken from the term structure. In this case the `forwardRate` is calculated as par coupon approximation from the term structure 
```cpp
            forwardRate_ =
                InterestRate((index_->forwardingTermStructure()->discount(valueDate_) /
                                  index_->forwardingTermStructure()->discount(maturityDate_) -
                              1.0) /
                                 index_->dayCounter().yearFraction(valueDate_, maturityDate_),
                             index_->dayCounter(), Simple, Once);
```
I am currently trying to understand how to fix this, but for the moment I am still missing the point where the test is logically failing. Any idea is highly appreciated. Thanks.

lballabio commented:
Ok, I got confused by the use of `QL_USE_INDEXED_COUPON` that seemed backwards, but it was just used to mean "the default case" and "the other case".  I changed the tests so that they run both cases regardless of that define, and I added a regression test for the failure you had (which depends on the date; it failed on February 23rd but it's passing now).

It won't seem like progress, since all the builds will fail now :)
But with the regression test in place, we can look at the dates and figure out the problem.

lballabio commented:
Ok, the problem is that the 3x6 FRA helper used for bootstrapping calculates the start date as `calendar.advance(spot, 3 months)` and the maturity date as `calendar.advance(spot, 6 months)` while the FRA instrument used for checking the curve calculates the start date in the same way but the maturity date as `index.maturityDate(start date)`, which basically means `calendar.advance(start date, 3 months)`.  In the regression case:

- today = Feb 23, 2023
- spot = Feb 27, 2023
- start date = May 29, 2023 (Feb 27 + 3 months, May 27 is Saturday, adjusted to 29)
- in the helper: maturity = Aug 28, 2023 (Feb 27 + 6 months, Aug 27 is Sunday, adjusted to 28)
- in the instrument: maturity = Aug 29, 2023 (May 29 + 3 months)

hence the difference when estimating the rate.

I think the implementation in the instrument makes more sense, and I would use it in the helper as well.  What do you people think?

lballabio commented:
Another thing I just realized: in the new constructor that takes the value and maturity dates, we probably need a forecast curve for the forward rate distinct from the optional discount curve which is only used for discounting the payoff.

jakeheke75 commented:
> I think the implementation in the instrument makes more sense, and I would use it in the helper as well. What do you people think?

It absolutely makes sense, in this way the two logic will be aligned.

> Another thing I just realized: in the new constructor that takes the value and maturity dates, we probably need a forecast curve for the forward rate distinct from the optional discount curve which is only used for discounting the payoff.

I apologize in advance, I am probably missing some important logic here. In the new constructor that takes the value and maturity date the discount curve should always be passed:
```cpp
    ForwardRateAgreement::ForwardRateAgreement(const Date& valueDate,
                                               const Date& maturityDate,
                                               Position::Type type,
                                               Rate strikeForwardRate,
                                               Real notionalAmount,
                                               Handle<YieldTermStructure> discountCurve,
                                               Natural fixingDays,
                                               BusinessDayConvention businessDayConvention)
```
In this way we are calculating the forward rate by using the rate specified in the `discountCurve` and we compare it with the contractualized `strikeForwardRate`. In attachment there is an example of a `discountCurve` built with with a `FlatForward` `YieldTermStructure`. Thanks for your time and patience.

[FRA_test.cpp.txt](https://github.com/lballabio/QuantLib/files/11133516/FRA_test.cpp.txt)

lballabio commented:
Sure, it's always passed&mdash;my point was that in the other constructors, the index refers to a forecast curve which is used to calculate the fixings, and the optional discount curve is used for discounting when given; if not, the forecast curve is also used for discounting.  Thus, it's possible to have separate curves for forecasting and discounting.  In the new constructor, the same curve is always used for both; which is ok as the default, but if we want to keep the possibility of a separate discounting curve we need the constructor to take a mandatory forecasting curve AND an optional discounting curve.

lballabio commented on ql/instruments/forwardrateagreement.cpp:
```diff
@@ -57,12 +58,67 @@ namespace QuantLib {
-    : ForwardRateAgreement(valueDate, index->maturityDate(valueDate), type, strikeForwardRate,
-                           notionalAmount, index, std::move(discountCurve), true) {}
+    : fraType_(type), notionalAmount_(notionalAmount),
+      useIndexedCoupon_(false), dayCounter_(discountCurve->dayCounter()),
```
Here's the problem — the discount curve might be empty.  Since we were taking the day counter from the index, I would probably add it as another parameter to the constructor, like the fixing days and the convention.  The same goes for the calendar, which was also taken from the index.

lballabio commented:
Can we take a step backwards, though, and check again what it is that we're trying to do?  I understand a forward contract on, say, Euribor3M at a given date.  But what it is that we're trying to model when passing two dates and no index?  Is it a bespoke contract to lend or borrow money at a given rate in the future?  If it is to be cash-settled, instead, who fixes the rate when the fixing date comes?

jakeheke75 commented:
Of course we can take a step back. I was reading again the original discussion of #154.
If I didn't misunderstand the discussion, the idea to have a costructor taking value date and maturity date but without an index is fulfilling the need of maximum flexibility in controlling the front end of the curve.
It should be a way to explicitly control the FRAs in a tighter fashion, therefore not in constant maturity space (as with the index) but at explicit fixing/value dates, as @mdelmedico was saying. Also @tomwhoiscontrary participated in the discussion, and @pcaspers as well. Your opinion is highly appreciated.

tomwhoiscontrary commented:
@jakeheke75 The more i have thought about this question, the less certain i have become! In particular, i think my initial comment on the bug is wrong, because i was conflating the index tenor (used to forecast a fixing) with the term of the notional loan underlying the FRA (used to calculate cashflows and for discounting). In practice these are the same, but by analogy with single-period swaps, i think the term of the loan could vary, and this would not affect the tenor used for the fixing.

I got interested in this because i was trying to use FRAs as a general tool for pulling arbitrary forward rates out of curves, and bootstrapping curves using arbitrary forward rates. The existing ForwardRateAgreement and FraRateHelper classes are not useful for that. But it now seems dubious that they *should* be useful for that. A FRA is not an arbitrary forward rate, it's a traded financial instrument defined in a specific way!

Given that it is an actual kind of instrument, it makes sense for the class to model the way the instrument actually works. Are we modelling an actual use with this change?

jakeheke75 commented:
Thanks @tomwhoiscontrary for sharing your point of view. Regarding the matter if we are modelling an actual use with this change... Well, it all started to have an "extra degree of flexibility". Yes, as you said, to have a control on forward rates not in constant maturity space (as with the index) but at explicit fixing/value dates. At this point, since I am not a professional working in the industry, I am not in the position to judge if this third constructor without the index will have a legit and effective practical application or not. If it is not useful, we have to remove it. We don't need useless functionalities. Will leave the decision to @lballabio and to @mdelmedico. Thanks everybody for participating in the discussion.

lballabio commented:
Well, I would definitely keep the part of this PR that deprecates the constructor taking both maturity date and index.  I would also keep the test for the regression you found while developing this and its eventual fix.

On the other hand, I would remove the new constructor without index and the new data members that supported it.  If we want to control the short end of the curve, I think it's better to be explicit and create a new helper for that.

lballabio commented:
I did some archaeology in the repo and found more info—I hope I can clear this up...

1) the calculation of the maturity date in the FRA helper (that is, as spot + 6 months for a 3x6 FRA, instead of as start + 3 months) was explicitly changed in https://github.com/lballabio/quantlib-old/pull/297.  As Paolo was working on an interest-rate desk at the time, I'd probably trust the change in absence of other info.

2) we also have https://github.com/lballabio/QuantLib/pull/706 by @pcaspers (ping, ping) which is relevant.  This is where we started having the option of calculating the FRA off the curve instead of asking the index (i.e., `useIndexedCoupon = false` or `true` respectively).

I'm not sure, but Peter's (2) may be the explanation of (1) as well: in the par approximation we're calculating start and maturity as spot + 3 months and spot + 6 months because they correspond to the dates of the corresponding coupon in a swap.

So, if we want to save the possibility to use the par approximation (we do), we need two things:

- in the rate helper, in the case `useIndexedCoupon = false`, keep the date calculation as it was because we do want that particular maturity.  (In the other case, we ask the index for the maturity.)

- in the instrument, we actually need a constructor that takes two dates if we want to calculate the same rate.  This doesn't mean that we want to pass arbitrary dates, but only that we want to be able to reproduce the par calculation.

Now, as for the constructor taking two dates: we want to pass two dates and a curve, or two dates and an index?  Given the above, I would probably still pass the index.  We're not modelling a FRA between two arbitrary dates: we're modelling a FRA on that index, but we're adjusting the dates to get the par approximation.

Another consequence is that the `useIndexedCoupon` argument in the constructor should probably disappear—but not the data member.  If we're using the constructor with start date and index, the maturity is necessarily calculated from the index, and therefore `useIndexedCoupon` must be `true` (even if we passed `false`, the maturity is the same).  Instead, if we're using the constructor with start and maturity dates, the only reason would be to use the par approximation, and therefore `useIndexedCoupon` must be `false` (if it were `true`, the passed maturity would be unused.)

If the above makes sense to you people, I'll try adding the corresponding changes to the PR.  Thoughts?

mdelmedico commented:
Remind me again why we would ever want to pass both start and end dates?
Wouldn’t you rather just pass the fixing or value date and get the rest of
the dates from the index since all the date math is correct via that
index?  I guess there is always the corner case that someone wants to pull
a 6m FRA off a 3m index, but that seems like a non-standard use case?


On Fri, Apr 21, 2023 at 07:56 Luigi Ballabio ***@***.***>
wrote:

> I did some archaeology in the repo and found more info—I hope I can clear
> this up...
>
>    1.
>
>    the calculation of the maturity date in the FRA helper (that is, as
>    spot + 6 months for a 3x6 FRA, instead of as start + 3 months) was
>    explicitly changed in lballabio/quantlib-old#297
>    <https://github.com/lballabio/quantlib-old/pull/297>. As Paolo was
>    working on an interest-rate desk at the time, I'd probably trust the change
>    in absence of other info.
>    2.
>
>    we also have #706 <https://github.com/lballabio/QuantLib/pull/706> by
>    @pcaspers <https://github.com/pcaspers> (ping, ping) which is
>    relevant. This is where we started having the option of calculating the FRA
>    off the curve instead of asking the index (i.e., useIndexedCoupon =
>    false or true respectively).
>
> I'm not sure, but Peter's (2) may be the explanation of (1) as well: in
> the par approximation we're calculating start and maturity as spot + 3
> months and spot + 6 months because they correspond to the dates of the
> corresponding coupon in a swap.
>
> So, if we want to save the possibility to use the par approximation (we
> do), we need two things:
>
>    -
>
>    in the rate helper, in the case useIndexedCoupon = false, keep the
>    date calculation as it was because we do want that particular maturity. (In
>    the other case, we ask the index for the maturity.)
>    -
>
>    in the instrument, we actually need a constructor that takes two dates
>    if we want to calculate the same rate. This doesn't mean that we want to
>    pass arbitrary dates, but only that we want to be able to reproduce the par
>    calculation.
>
> Now, as for the constructor taking two dates: we want to pass two dates
> and a curve, or two dates and an index? Given the above, I would probably
> still pass the index. We're not modelling a FRA between two arbitrary
> dates: we're modelling a FRA on that index, but we're adjusting the dates
> to get the par approximation.
>
> Another consequence is that the useIndexedCoupon argument in the
> constructor should probably disappear—but not the data member. If we're
> using the constructor with start date and index, the maturity is
> necessarily calculated from the index, and therefore useIndexedCoupon
> must be true (even if we passed false, the maturity is the same).
> Instead, if we're using the constructor with start and maturity dates, the
> only reason would be to use the par approximation, and therefore
> useIndexedCoupon must be false (if it were true, the passed maturity
> would be unused.)
>
> If the above makes sense to you people, I'll try adding the corresponding
> changes to the PR. Thoughts?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lballabio/QuantLib/pull/1600#issuecomment-1517796848>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AH64OZV3VXICY22PZGNYAT3XCJ7ZPANCNFSM6AAAAAAUY7OOVU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

lballabio commented:
See https://github.com/lballabio/QuantLib/pull/706

lballabio commented:
@pcaspers, maybe you can elaborate a bit on what you wrote in the PR I linked?

jakeheke75 commented:
> Another consequence is that the `useIndexedCoupon` argument in the constructor should probably disappear—but not the data member. If we're using the constructor with start date and index, the maturity is necessarily calculated from the index, and therefore `useIndexedCoupon` must be true (even if we passed false, the maturity is the same). Instead, if we're using the constructor with start and maturity dates, the only reason would be to use the par approximation, and therefore `useIndexedCoupon` must be false (if it were true, the passed maturity would be unused.)

OK Luigi, which do you think is the best way to avoid ambiguity in the different constructors? The first constructor that we deprecated is taking the `valueDate`, `maturityDate` and the `useIndexedCoupon` argument with default value `true`. According to the logic that we want to implement, in this case this is not correct and it must be always `false`. But I think that we shouldn't modify this first constructor right? Otherwise we are probably going to break existing user code.
So let's say that we keep this first costructor as deprecated (even if it is probably a bit inconsistent), and then we have the second one that will always have `useIndexedCoupon = true`:
```cpp
        ForwardRateAgreement(
            const Date& valueDate,
            Position::Type type,
            Rate strikeForwardRate,
            Real notionalAmount,
            const ext::shared_ptr<IborIndex>& index,
            Handle<YieldTermStructure> discountCurve = Handle<YieldTermStructure>());
```
We create a third one that will always have `useIndexedCoupon = false`:
```cpp
        ForwardRateAgreement(
            const Date& valueDate,
            const Date& maturityDate,
            Position::Type type,
            Rate strikeForwardRate,
            Real notionalAmount,
            const ext::shared_ptr<IborIndex>& index,
            Handle<YieldTermStructure> discountCurve = Handle<YieldTermStructure>());
```
In order to avoid ambiguity between the first deprecated constructor and the new third one, do we have to change the order of the arguments, even if we are somehow breaking the logical sequence? Something like:
```cpp
        ForwardRateAgreement(
            const Date& valueDate,
            Position::Type type,
            Rate strikeForwardRate,
            Real notionalAmount,
            const Date& maturityDate,
            const ext::shared_ptr<IborIndex>& index,
            Handle<YieldTermStructure> discountCurve = Handle<YieldTermStructure>());
```
Thanks

lballabio commented:
I think we can deprecate both existing constructors and add new ones for the two cases we want.  To disambiguate, we can for instance move the index to the first place in the new constructors.

tomwhoiscontrary commented:
> I think we can deprecate both existing constructors and add new ones for the two cases we want.

So there will be a constructor taking only a start date, which creates a FRA using indexed coupons, and one taking both start and end dates, which creates a FRA using par coupons?

So the choice between par and indexed coupons will to be made implicitly by whether you pass in an end date or not? On the one hand, it seems wrong that it's not explicit. On the other hand, if i understand correctly, "par coupons" just means using a custom (incorrect 😀) end date for the forward rate, so maybe what is being implied is no more than what is explicit.

Actually, can i check my understanding? I have never really got a good grip on this "par coupon" stuff. Does it just mean that when we are forecasting a rate for a coupon, we use the forward rate over the accrual period of the coupon, rather than over the accrual period of the notional deposit underlying the index?

If so, i think a par coupon is exactly what i was looking for all those years ago when i first encountered this issue.

But still, the proposed design does mean that you can express a par coupon FRA with a non-standard tenor (eg 3x5 on 3M Euribor), but cannot express an indexed coupon FRA with such a tenor. Do we want that? If there is a case for expressing such FRAs, why would it be restricted to the par coupon case?

On the other hand, if the only motivation for taking an end date is allowing par coupons, then aren't we forcing the user to explicitly work out the end date themselves if they want to do that? Wouldn't it be better to work out the accrual period in the constructor?

My opinion today (which will probably be different tomorrow, sorry), is that two designs are internally consistent:

1. one constructor, taking a start date, no end date, and a flag for par vs indexed coupons; the accrual end date is worked out using the index tenor, and the rate used is either indexed (forward over the index tenor) or par (forward over the accrual period)

2. two constructors, the above, plus a constructor taking a start date and an end date, and a flag for par vs indexed coupons; we use the specified end date for the end of the accrual period, then determine the rate according to the flag as above.

Is it worth thinking about adding a MakeForwardRateAgreement, to give more flexibility and explicitness here, while allowing ForwardRateAgreement to have a single kitchen-sink constructor?

lballabio commented:
> Does it just mean that when we are forecasting a rate for a coupon, we use the forward rate over the accrual period of the coupon, rather than over the accrual period of the notional deposit underlying the index?

Right.  The problem is that the information about the end date must be passed from the outside; it can't be deduced from the start date, because it depends on the larger schedule of which the coupon is a part.

tomwhoiscontrary commented:
> Right. The problem is that the information about the end date must be passed from the outside; it can't be deduced from the start date, because it depends on the larger schedule of which the coupon is a part.

Oh, even for FRAs? I understand that for swap coupons, but i assumed a FRA was a standalone thing.

lballabio commented:
True, but still it can't be deduced from the start date alone.  For a NxM FRA, the "par" start and end dates are d0 + N months and d0 + M months, where d0 is the creation date of the FRA.

If we only have the start date (d0 + N months), we can't deduce d0 + M months because (a) we don't know what N is; from the index, we can only get its tenor = M-N; and (b) even if we knew N, we can't deduce d0.  If the start date is, say, May 8th 2023 and NxM=3x6, d0 might have been February 8th, or it might have been February 6th which also gives May 8th as the value date because May 6th is a Saturday.  If we can't know that, we can't know if the "par" maturity date should be August 7th (the 6th is a Sunday) or August 8th.
